bin/create_release_archives.sh

```diff
@@ -0,0 +1,119 @@
+
+set -o errexit
+set -o pipefail
+set -x
+
+# This script primarily exists for Cloud Builder.  This script
+# reads artifacts from a specified directory, generates tar files
+# based on those artifacts, and then stores the tar files
+# back to the directory.
+
+BASE_DIR="$(mktemp -d /tmp/istio.version.XXXX)"
+OUTPUT_PATH=""
+VER_STRING=""
+
+function usage() {
+  echo "$0
+    -d <path> path to use for temp directory                  (optional, randomized default is ${BASE_DIR} )
+    -o <path> path where build output/artifacts are stored    (required)
+    -v <ver>  version info to include in filename (e.g., 1.0) (required)"
+  exit 1
+}
+
+function error_exit() {
+  # ${BASH_SOURCE[1]} is the file name of the caller.
+  echo "${BASH_SOURCE[1]}: line ${BASH_LINENO[0]}: ${1:-Unknown Error.} (exit ${2:-1})" 1>&2
+  exit ${2:-1}
+}
+
+while getopts d:o:v: arg ; do
+  case "${arg}" in
+    d) BASE_DIR="${OPTARG}";;
+    o) OUTPUT_PATH="${OPTARG}";;
+    v) VER_STRING="${OPTARG}";;
+    *) usage;;
+  esac
+done
+
+[[ -z "${BASE_DIR}"  ]] && usage
+[[ -z "${OUTPUT_PATH}"  ]] && usage
+[[ -z "${VER_STRING}"   ]] && usage
+
+COMMON_FILES_DIR="${BASE_DIR}/istio/istio-${VER_STRING}"
+BIN_DIR="${COMMON_FILES_DIR}/bin"
+mkdir -p "${BIN_DIR}"
+
+# On mac, brew install gnu-tar gnu-cp
+# and set CP=gcp TAR=gtar
+
+if [[ -z "${CP}" ]] ; then
+  CP=cp
+fi
+if [[ -z "${TAR}" ]] ; then
+  TAR=tar
+fi
+
+function create_linux_archive() {
+  local istioctl_path="${BIN_DIR}/istioctl"
+
+  ${CP} "${OUTPUT_PATH}/istioctl-linux" "${istioctl_path}"
+  chmod 755 "${istioctl_path}"
+
+  ${TAR} --owner releng --group releng -czvf \
+    "${OUTPUT_PATH}/istio-${VER_STRING}-linux.tar.gz" "istio-${VER_STRING}" \
+    || error_exit 'Could not create linux archive'
+  rm "${istioctl_path}"
+}
+
+function create_osx_archive() {
+  local istioctl_path="${BIN_DIR}/istioctl"
+
+  ${CP} "${OUTPUT_PATH}/istioctl-osx" "${istioctl_path}"
+  chmod 755 "${istioctl_path}"
+
+  ${TAR} --owner releng --group releng -czvf \
+    "${OUTPUT_PATH}/istio-${VER_STRING}-osx.tar.gz" "istio-${VER_STRING}" \
+    || error_exit 'Could not create osx archive'
+  rm "${istioctl_path}"
+}
+
+function create_windows_archive() {
+  local istioctl_path="${BIN_DIR}/istioctl.exe"
+
+  ${CP} "${OUTPUT_PATH}/istioctl-win.exe" "${istioctl_path}"
+  
+  zip -r "${OUTPUT_PATH}/istio_${VER_STRING}_win.zip" "istio-${VER_STRING}" \
+    || error_exit 'Could not create windows archive'
+  rm "${istioctl_path}"
+}
+
+pushd "${OUTPUT_PATH}"
+${CP} istio.VERSION LICENSE README.md CONTRIBUTING.md "${COMMON_FILES_DIR}"/
+find samples install -type f \( -name "*.yaml" -o -name "cleanup*" -o -name "*.md" \) \
+  -exec ${CP} --parents {} "${COMMON_FILES_DIR}" \;
+find install/tools -type f -exec ${CP} --parents {} "${COMMON_FILES_DIR}" \;
+popd
+
+# Changing dir such that tar and zip files are
+# created with right hiereachy
+pushd "${COMMON_FILES_DIR}/.."
+create_linux_archive
+create_osx_archive
+create_windows_archive
```

Comment:
Can you create makefile targets for these actions?

Reply:
Currently there's not a decent way to have a separate target for each OS.  Each archive only differs by the istioctl executable (and each of these 3 are made together by upload-istioctl [which uses "bazel build" for Linux and then [with help from bazel_to_go] uses "go build" for Mac and Windows -- I'm just the messenger, not the inventor, regarding this).

Even with one target I'm not clear on where the resulting archives should go since Istio doesn't yet seem to have an established & centralized output directory in make.  Everything else in the Makefile either just runs bazel (which has established output directories like bazel-bin/, which seems odd to have make use [and not very future proof is people are serious on getting ride of bazel] ), and the other make targets just push to docker hubs.  The create_release_archives.sh script basically relies on cloud_builder.sh to assemble everything together in one local directory, and both are using an output directory that's passed in on the command-line.

Otherwise, the tar files currently just use istioctl (from pilot) plus other static content from the same repo so in that sense a combined make target is feasible but this will become a bigger challenge to maintain if/when the archives start to provide more diverse content (e.g., from proxy or other repos).  For I kept cloud_build.sh and create_release_archives.sh separate to make a clear boundary between compile and installer steps (and provide a more extensible opportunity to run other steps [3rd party compile in another docker image] between those 2 steps.

Reply:
I think it would be good to reunite the green build with this since they are supposed to be the same, the only difference being one is incremental while the other is clean. Right now you might have a wrong signal from the green build.

Reply:
It would be good to eventually bring them closer together.  I've already experienced a wrong signal since after updating to the latest istio I hit an issue that requires the following fix even though the commit I updated to was okayed by green build:

https://github.com/istio/istio/pull/1716

Reply:
You can create the artifacts directory for each repo, and have something like ARTIFACTS_DIR="" make artifacts. I think the fact that we use muliple building is the reason to have a common Makefile such that developer don't have to look at the code to know how to build. https://github.com/istio/istio/blob/master/DEV-GUIDE.md#building-the-code